pkg/importer/dataStream.go

```diff
@@ -915,51 +845,38 @@ func (d *DataStream) convertQcow2ToRaw(src, dest string) error {
 }
 
 // Copy endpoint to dest based on passed-in reader.
-func (d *DataStream) copy(dest string) (err error) {
+// if the error is not nil and true is return, we require scratch space to complete the copy.
+func (d *DataStream) copy(dest string, hasScratch bool, scratchDataDir string) (bool, error) {
 	if d.isHTTPQcow2() {
-		err = d.convertQcow2ToRawStream(dest)
+		err := d.convertQcow2ToRawStream(dest)
 		if err != nil {
-			return err
+			return false, err
 		}
 	} else {
-		dest = filepath.Clean(dest)
-		glog.V(2).Infof("copying image file to %q", dest)
-		tmpDest := dest
-		if !d.isQemuStored() {
-			//it is either not qemu or not stored
-			if d.qemu {
-				// copy to tmp; qemu conversion will write to passed-in destination
-				tmpDest = randTmpName(dest)
-				glog.V(3).Infof("Copy: temp file for qcow2 conversion: %q", tmpDest)
-				defer func(f string) {
-					os.Remove(f)
-				}(tmpDest)
-			}
-			if d.isMoveable() {
-				glog.V(3).Infof("Moving already stored raw from %s to %s", d.tmpDataPath, tmpDest)
-				err = MoveFile(d.tmpDataPath, tmpDest)
-				if err != nil {
-					glog.V(3).Infof("Failing back to copying stored raw from %s to %s", d.tmpDataPath, tmpDest)
-					//Failback to copy since move failed
-					err = StreamDataToFile(d.topReader(), tmpDest)
-				}
-			} else {
-				glog.V(3).Infof("Streaming data of non Movble file to %s", tmpDest)
-				err = StreamDataToFile(d.topReader(), tmpDest)
-			}
-			if err != nil {
-				return errors.WithMessage(err, fmt.Sprintf("unable to stream data to file %q", dest))
-			}
-		} else {
-			//It is a stored qemu file - taking a location where it is stored
-			glog.V(3).Infof("Streaming file to %s", tmpDest)
-			tmpDest = d.tmpDataPath
+		if !hasScratch {
+			//Need scratch space but none provided.
+			return true, errors.New("Scratch space required, and none found")
+		}
+		// Replace /data/target name with scratch path/target name
+		tmpDest := filepath.Join(scratchDataDir, filepath.Base(dest))
+
+		err := StreamDataToFile(d.topReader(), tmpDest)
+		if err != nil {
+			return false, err
+		}
+
+		// Make sure the destination is empty
+		glog.V(3).Infoln("Cleaning up my mess.")
+		err = cleanDir(d.DataDir)
```

Comment:
The problem is you return the reader to data stream, and the reader is just an open file to the unconverted disk image, but the copy code tries to write that reader to the temporary space, but its already there with the same name so we can't do that.

I don't like the way its solved right now by some flags that say don't copy or don't move. That gives a very complex flow that makes it hard to extend in the future. It should be fairly simple:

if canStreamConvert {
  do streaming conversion
} else {
 write contents of reader to temporary location
  do non streaming conversion (although technically the difference between streaming and non streaming is the intermediate write to scratch, so we could write first, and then alter the url to a file:///path, and do the conversion from there)
}
resize if needed.

Reply:
What we have prior to scratch space enhancement is:
1. registry()
2. create directory /data/tmp
3. download container layers to /data/tmp/datatmp
4. extract only disk directory to /data/tmp/
5. obtain file reader on /data/tmp/imageifle
6. copy :
7. In case the file is qcow2 - do not copy to tmp - since it is already Stored -> just convert o raw and store converted under /data/disk.img
8. In case the file is raw - move it (this is an optimization inside the same fs, itr may fail if src and dst are not in the same fs -> failback to copy) from, /data/tmp/disk/imagefile to /data/disk.img
9. In all other cases (archived qcow2 etc.) - copy the file to /data/tmp<mangling>/ and then convert to raw and store in /data/disk.img
10. On ds.Close() call Clenaup of dataCtxt received from registry() flow - it will cleanuo the entire /data/tmp directory

So in fact, the entire /data/tmp dircetory and all that is there can be siwtched to be on scrtach space - this is a minor change of TempContainerDiskDir. 

I agree that when working with scratch space - there is no need in Moveable() is isStored() etc. If we can afford that we should keep it simple:
All tmp data on scratch space and only the resulting converted disk.img on /data/disk.img